GCS: Launch setup wizard when unconfigured board is connected

tracernz opened the pull request:
Got bored, wrote a little bit of code. This __always__ pops the setup wizard when a board without actuators configured is connected. Thoughts?

"unconfigured" means no non-zero channels in ActuatorSettings.

Fixes #1737.

mlyle commented:
Holy crap that was fast.

My one thought is--- how does this play with the auto-upgrade wizard?  It would be nice if A) the auto-upgrade wizard's dialog always "won" and was presented first, so if someone plugs in an old seppuku and then starts GCS they don't get two weird things triggered (I understand the right thing probably happens if we catch the booatloader.  and B) if we presented a similar dialog for this and allowed a similar "ignore" to take place.

tracernz commented:
The auto-upgrade dialog wins if you plug the board in when GCS is running because it happens at the bootloader. This fires once the telemetry connection is actually established and the settings (and meta, and update-on-change) objects are all fetched. I will take a look at how it will play out if GCS is started when the board is already running+plugged-in. 

The reason I didn't want to add another dialog is because I didn't want to add yet another thing to click through. I agree it would be good so I propose to add an ignore button on the first screen of the setup wizard (but only if it was launched automatically). Do you think that's a good solution?

-e- Also need to verify how it plays with the upgrader when it boots firmware to import settings!

mlyle commented:
> The reason I didn't want to add another dialog is because I didn't want to add yet another thing to click through. I agree it would be good so I propose to add an ignore button on the first screen of the setup wizard (but only if it was launched automatically). Do you think that's a good solution?

That sounds great.

This is a bit of a tangent, but I'm thinking about replacing the 'Rescue' button / timer with a pulldown that is "On board connection:" 'Stop in bootloader' / 'Prompt for upgrades' / 'Immediately boot'.  Immediately boot would probably also bypass the post-boot firmware prompting.  Thoughts?

tracernz commented:
Just to keep this up-to-date, I have some WIP on my laptop to play nice with upgrader etc. Hoping to make upgrader offer setup wizard if no settings are imported (would fix #1697 too). Am out of town next week so likely to spend a bit of time on my laptop. ;)

tracernz committed:
GCS:SetupWizard: Don't auto-launch during upgrader

Monkey patched due to state of uploader plugin. Needs fixed later when stuff isn't all done in the UAVGadget!

tracernz commented:
This now only pops setup wizard if the firmware version matches GCS (in which case upgrader doesn't pop). Monkey-patched it so that it won't pop after the upgrader boots the new firmware too (very icky with current state of uploader). Doesn't go to setup wizard after upgrade if the board is unconfigured.

mlyle commented:
Sorry, has conflicts, needs rebase :/

mlyle committed:
GCS: Launch setup wizard when unconfigured board is connected

"unconfigured" means no non-zero channels in ActuatorSettings

mlyle commented:
Hey look I'm allowed to rebase

mlyle commented on ground/gcs/src/plugins/config/configplugin.cpp:
```diff
@@ -75,9 +75,9 @@ bool ConfigPlugin::initialize(const QStringList &args, QString *errMsg)
     connect(telMngr, &TelemetryManager::connected, this, &ConfigPlugin::onAutopilotConnect);
     connect(telMngr, &TelemetryManager::disconnected, this, &ConfigPlugin::onAutopilotDisconnect);
 
-    // And check whether by any chance we are not already connected
```
So we're positive that initialization of all plugins (and all connections made) before autopilot connection?  I know it's generally slow/takes a few seconds.. I'm just curious if this is guaranteed by design.

tracernz replied:
We don't have a device list until after all the plugins are loaded: https://github.com/d-ronin/dRonin/blob/next/ground/gcs/src/plugins/coreplugin/connectionmanager.cpp#L466

mlyle commented on ground/gcs/src/plugins/setupwizard/setupwizard.cpp:
```diff
@@ -366,3 +368,25 @@ bool SetupWizard::saveHardwareSettings() const
+        return;
+
+    QMessageBox prompt(QMessageBox::Question, tr("Ignore Board?"),
+                       tr("Would you like to ignore the setup wizard for this board?"),
```
This should be something like.. "This board doesn't appear to be configured.  Would you like to automatically configure it now with the dRonin Setup Wizard?"

tracernz replied:
I'll take your wording in your later comment and apply it later today. :+1:

mlyle commented on ground/gcs/src/plugins/setupwizard/setupwizard.cpp:
```diff
@@ -366,3 +368,25 @@ bool SetupWizard::saveHardwareSettings() const
+    QMessageBox prompt(QMessageBox::Question, tr("Ignore Board?"),
+                       tr("Would you like to ignore the setup wizard for this board?"),
+                       QMessageBox::Yes | QMessageBox::No);
+    if (prompt.exec() == QMessageBox::No)
```
(and of course invert the sense of this if that wording is used)

mlyle reviewed (changes requested):
Looks good overall.  I think the dialog wording is the only thing that needs to change.

mlyle commented:
OK, I tested this and hadn't quite understood the dialog-- it comes up on cancel.

Maybe something like "dRonin Setup Wizard Cancelled!"   "Would you like to prevent the Setup Wizard from automatically opening for this board? (Only applies for this session)."

Otherwise, tested a few cases, seems good :)

tracernz committed:
GCS:SetupWizard: Tweak wording for auto ignore dialog

tracernz commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

tracernz commented:
jenkins, build artifacts please

and bring me a shrubbery

tracernz commented:
Tested on Win 10 with Seppuku (incl. upgrade). No problems.
